0xProject/0x-monorepo
# Refactor Blockchain.ts to allow arbitrary provider state changes

## Description
This started off as a task to support the new metamask (which will stop refreshing on network change https://github.com/MetaMask/metamask-extension/pull/4558/files#diff-d0dadea64ae305b53b5c625db0e0ef1fL28) but then I realized that the architecture doesn't support changing network ID (although there was some code that implied it could, what was that trying to achieve, anyone know?) so I refactored things a bit.

Supporting the new metamask involves using a somewhat unofficial (?) api where you can subscribe to changes in metamask.

The refactor involves just having one function when we want to init providers / web3wrappers or when we want to change them (such as on network change, or using the ledger).

## Testing instructions
You need to download the "beta" version of metamask.
https://twitter.com/fragosti/status/1008875873917808645

And then access Portal and Legacy Portal and try going between networks and providers to make sure everything works as expected. 
Also you should try both versions of metamask.

I've done all this already, including testing / fixing ledger support.
## Types of changes




## Checklist:
